Allow descriptor imports with importmulti

meshcollider opened the pull request:
~~Based on #14454 #14565, last two commits only are for review.~~

Best reviewed with `?w=1`

Allows a descriptor to be imported into the wallet using `importmulti` RPC. Start and end of range can be specified for ranged descriptors. The descriptor is implicitly converted to old structures on import.

Also adds a simple test of a P2SH-P2WPKH address being imported as a descriptor. More tests to come, as well as release notes.

sipa commented on src/wallet/rpcdump.cpp:
```diff
@@ -806,266 +807,338 @@ UniValue dumpwallet(const JSONRPCRequest& request)
+        if (::IsMine(*pwallet, raw_pubkey_script) == ISMINE_SPENDABLE) {
+            throw JSONRPCError(RPC_WALLET_ERROR, "The wallet already contains the private key for this public key");
+        }
+        if (!pwallet->AddWatchOnly(raw_pubkey_script, timestamp)) {
```
For the same reason as pointed out in the other PR, you can't do this; you're going to mark payments to individual multisig pubkeys as incoming payments.

instagibbs replied:
Restricting it from multisig(to avoid being tricked as you mention) would make this even more confusing to a user. 

Unfortunate.

MeshCollider replied:
Mentioned on IRC, but what if we just never import public keys at all, and only either A) add the scriotPubKey as watch only or B) import the private key. If we have the private key then IMO it's less scary, because it's still "ours" and we can access the funds

sipa replied:
You're right; the same concern doesn't exist for private keys as you're obviously able to spend those coins anyway.

instagibbs replied:
Would not importing the pubkeys affect the wallet's ability to construct PSBT inputs?

sipa replied:
Only for things that have PKH/WPKH construction in them (GetPubKey is needed for those, which finds a pubkey based on pubkeyhash).

promag reviewed (commented):
Just some comments while reading the code. Will test.

MeshCollider commented on src/wallet/rpcdump.cpp:
```diff
@@ -806,266 +807,338 @@ UniValue dumpwallet(const JSONRPCRequest& request)
+        } else if (data.exists("descriptor")) {
+            ProcessImportDesc(pwallet, data, timestamp);
+        } else {
+            throw JSONRPCError(RPC_INVALID_PARAMETER, "Either a descriptor or scriptPubKey must be provided.");
```
I said more tests were coming 😅

Sjors replied:
Still coming?

instagibbs replied:
looks like it came: https://github.com/bitcoin/bitcoin/pull/14491/commits/fbb5e935eaf17d603ec62e1a704a174235540b71#diff-3dfbfa462305488434b8d8da81f99de7R613

MeshCollider commented:
After discussion with sipa, closing for now, will come back to this after #14454 is merged

achow101 commented:
Since #14454 has been merged, this can be reopened?

sipa commented:
I believe a simpler approach is possible on top of #14565.

Sjors commented on src/wallet/rpcdump.cpp:
```diff
@@ -1133,6 +1207,7 @@ UniValue importmulti(const JSONRPCRequest& mainRequest)
             "      \"witnessscript\": \"<script>\"                           , (string, optional) Allowed only if the scriptPubKey is a P2SH-P2WSH or P2WSH address/scriptPubKey\n"
             "      \"pubkeys\": [\"<pubKey>\", ... ]                         , (array, optional) Array of strings giving pubkeys that must occur in the output or redeemscript\n"
             "      \"keys\": [\"<key>\", ... ]                               , (array, optional) Array of strings giving private keys whose corresponding public keys must occur in the output or redeemscript\n"
+            "      \"range\": [<start>, <end>]                               , (array, optional) If a ranged descriptor is used, this specifies the start and end of the range to import\n"
```
Mention default (standard key pool size?)

Sjors reviewed (commented):
Concept ACK, lightly tested 6894bf5 and it's able to import a test watch-only wallet I used previously.

I didn't use the `watchonly` while importing a descriptor with an `xpub`, should that be allowed? Or should `watchonly` argument not be allowed when using a descriptor?

Much smaller code change than I would have thought, nice! 

It would be nice if `range` had a default of `[0, DEFAULT_KEYPOOL_SIZE]` or even using the `-keypool` argument.

Can you add a usage example with a descriptor?

When I cherry-pick #14477 on top of this and inspect a used address with `getaddressinfo`, the origin info in the descriptor seems both wrong and incomplete: `"desc": "wpkh([224885f8]026...)",` where `224885f8` is not the master fingerprint I used, and derivation info is missing. There might be some reusable stuff in #14021.

If I use `watchonly: true`, then `getaddressinfo` doesn't show a descriptor and it says `solvable: false`, which seems wrong (the latter also happens without cherry-pick).

sipa commented:
@MeshCollider @Sjors I was imagining that "watchonly" would be implicit when using descriptors (`addr()` and `raw()` would be watchonly; anything else would result in a solvable result).

The reason for "watchonly" is so that users need to be explicit about the fact they _don't_ want solvability (generally, you should always want solvability, but if you can't, you can tell importmulti that it's fine without).

Sjors commented:
@sipa I like your suggestion. In that case we should disallow usage of the `watchonly` param when combined with a descriptor.

Regarding getting origin information from the descriptors, @achow101  just rebased #14021 on top of this PR. It's a non trivial change. Perhaps for this PR it's best to not store origin information. Just make sure that if you do `walletcreatefundedpsbt` with bip32 flag set to true, then the result doesn't crash `decodepsbt`.

MeshCollider commented:
I don't see the problem if you want to add a ranged descriptor as watch only to import all the scriptPubKeys but not retain any more info than that?

Sjors commented:
IIUC you would indicate that in the descriptor by wrapping the result in `addr(...)`. Mandating that removes ambiguity from how a descriptor ends up in a wallet.

MeshCollider commented:
@Sjors It was my understanding that `addr(...)` cannot contain another type of descriptor or key (e.g. a ranged BIP32 key), only the base58/bech32 encoded address formats

instagibbs commented:
xpub byte prefix mismatch results in very generic error; would be nice to give something more meaningful

meshcollider committed:
Add release notes for importmulti descriptor support

practicalswift commented on src/wallet/rpcdump.cpp:
```diff
@@ -804,233 +805,351 @@ UniValue dumpwallet(const JSONRPCRequest& request)
+            // RecurseImportData() removes any relevant redeemscript/witnessscript from import_data, so we can use that to discover if a superfluous one was provided.
+            if (import_data.redeemscript) warnings.push_back("Ignoring redeemscript " + strRedeemScript + " as this is not a P2SH script.");
+            if (import_data.witnessscript) warnings.push_back("Ignoring witnessscript " + witness_script_hex + " as this is not a P2WSH script.");
+            for (const auto& privkey : import_data.privkeys) {
```
Could use `std::any_of`? :-)

instagibbs commented:
Can you rebase this on most recent #14565 ? It has a few logic fixes.

In addition, you can not ignore the `watch_only` field for descriptor import. `watch_only` needs to be set to true if the descriptor deals with public keys only: https://github.com/bitcoin/bitcoin/pull/14565/files#diff-522490d83dce5375d423b23886e4125eR1019

Sjors commented:
Now might be a good to time to rebase it, because testing this might reveal anything we've missed in the overhaul PR #14565.

meshcollider committed:
[wallet] Allow descriptor imports with importmulti

meshcollider committed:
Add test for importing via descriptor

MeshCollider commented:
I've rebased this but haven't tested, it was a bit of a messy rebase and still a couple of things to address such as adding the warnings to the descriptor import function, plus more tests. I'll hopefully finish those things off in the next couple of days too.

Sjors commented:
By "haven't tested" you mean "haven't compiled" right? :-)

It looks like a tricky rebase. E.g. your code is referring to `import_data.pubkeys` and `import_data.privkeys`, the latter has been renamed to `import_data.used_keys`, but the former doesn't exist anymore.

The `ProcessImport` function is now duplicated.

MeshCollider commented:
I've now finished rebasing, lightly tested, added some warnings, and added a couple more tests. There are a couple of TODO's left in the code which I will come back to

achow101 commented on src/wallet/rpcdump.cpp:
```diff
@@ -923,203 +924,291 @@ static std::string RecurseImportData(const CScript& script, ImportData& import_d
+static UniValue ProcessImport(CWallet * const pwallet, const UniValue& data, const int64_t timestamp) EXCLUSIVE_LOCKS_REQUIRED(pwallet->cs_wallet)
+{
+    try {
+        bool watch_only = data.exists("watchonly") ? data["watchonly"].get_bool() : false;
```
This variable is unused in this function

jnewbery commented:
Concept ACK. I've had a quick skim of the code. A couple of high-level questions:

- have you considered implementing this as a separate `importdescriptor` RPC method? There are enough modal arguments (eg redeemscript/witnessscript/pubkeys/watchonly(?) should only be used when importing addresses/scripts, range should only be used when importing descriptors) that it seems like this makes sense as a separate method.
- should it ever be necessary to add a `keys` argument for importing private keys with a descriptor? Can't you just include the private keys directly in the descriptor?

instagibbs commented:
@jnewbery IIRC a unique RPC would likely be reserved for when wallet actually has descriptor records and can have a clean API break.

MeshCollider commented:
@jnewbery @Sjors as mentioned above

>There is one thing that still needs addressing, which is that a descriptor which provides a private key should import that private key. Currently it only Expands() the descriptor and imports the public keys. The implementation of that is likely going to require some discussion so I will open a future PR for that, and leave it as a TODO for now.

So right now, the "keys" field is the only way to import a private key. That is because the descriptor code doesn't currently support deriving specific private keys, Expand() only provides public keys and scripts. I have the code for importing private keys too ready, so let's discuss that issue there when I open a PR.

MeshCollider commented:
Rebased

instagibbs reviewed (commented):
Suggestion though not your fault: change`Parse` to `ParseDescriptor`?

Sjors commented on src/wallet/rpcdump.cpp:
```diff
@@ -1108,14 +1117,25 @@ static UniValue ProcessImportDescriptor(ImportData& import_data, std::map<CKeyID
         // Check if this private key corresponds to a public key from the descriptor
         if (!pubkey_map.count(id)) {
-            throw JSONRPCError(RPC_INVALID_ADDRESS_OR_KEY, "Unused private key provided");
+            warnings.push_back("Ignoring irrelevant private key.");
```
This could use a test.

Sjors commented on src/wallet/rpcdump.cpp:
```diff
@@ -1108,14 +1117,25 @@ static UniValue ProcessImportDescriptor(ImportData& import_data, std::map<CKeyID
+    // This does not take into account threshold multisigs which could be spendable without all keys
+    bool spendable = std::all_of(pubkey_map.begin(), pubkey_map.end(), [&](const std::pair<CKeyID, CPubKey>& used_key){ return privkey_map.count(used_key.first) > 0; });
+    if (!watch_only && !spendable) {
+        warnings.push_back("Some private keys are missing, outputs will be considered watchonly. If this is intentional, specify the watchonly flag.");
```
Also, this and the related warning could use a test.

Sjors reviewed (commented):
For readability it might help if you split 9738518cf2e09cc77f252ccf2bc7ee2d938ed5c9 into:
* one commit where you introduce `ProcessImportLegacy`
* one where you refactor it (it's probably just a few changes, but GitHub is confused by it)
* one where you add descriptor support.

I'd also like to see slightly more tests, e.g. one with a multisig descriptor.

promag commented:
> Would it be better to only allow importing privkeys from the descriptor itself?

@jnewbery you mean deprecate `keys` in `importmulti` or error if `keys` and `desc` are used?

jnewbery commented:
> you mean deprecate keys in importmulti or error if keys and desc are used?

We can't deprecate `keys` in `importmulti`. That field is still required if importing a script not using a descriptor. My suggestion would be to error if `desc` and `keys` are provided in the same import object (in the same way that this PR errors if both `desc` and `scriptPubKey` are provided).

But perhaps I don't have full context, and importing with a descriptor and private keys provided separately is important functionality. Looking forward to hearing meshcollider's input.

sipa commented:
@jnewbery I think it's reasonable to have a shared fully-public descriptor that all participants know, and each has their own private key stored separately. It may be inconvenient to first need a utility to combine the public descriptor with private keys before importing.

MeshCollider commented:
Agree with sipa, I don't see a reason for limiting one way or the other just for the sake of limitation, this flexibility seems potentially useful unless there's a case against it

jnewbery reviewed (commented):
I think there are still some residual issues from the rebase (see my comment on the `warnings` and `result` handling).

I agree with @Sjors comment here: https://github.com/bitcoin/bitcoin/pull/14491#pullrequestreview-189838451 that this would be *much* easier to review if you re-arranged the commits. The first commit is a combination of code-move (from `ProcessImport()` to `ProcessImportLegacy()`); refactor of `ProcessImport()`; and new functionality. If you split that into 3 or more commits, it'd be easier for reviewers. I'd also suggest squashing the following into earlier commits once that's done:

- 5330527724c9af9e79eafc6c5fc9df62e7695161 (*Add warnings to descriptor imports based on watchonly*)
- d2fe87244d64c5cfadf108c570bd50ca8139e239 (*Improve descriptor import tests for importmulti*)
- 63b669bf4d9c6effa5465ed2e843fb7d61e5a2da (*Remove unused watch_only variable from ProcessImport*)

Happy to help with some of that if you need a hand.

sipa commented on src/wallet/rpcdump.cpp:
```diff
@@ -923,154 +924,251 @@ static std::string RecurseImportData(const CScript& script, ImportData& import_d
+        import_data.import_scripts.emplace(x.second);
+    }
+
+    std::copy(out_keys.pubkeys.begin(), out_keys.pubkeys.end(), std::inserter(pubkey_map, pubkey_map.end()));
```
I don't think this is correct, as it will cause import of all involved public keys as watch-only (even for example the public keys that are part of a multisig).

MeshCollider replied:
Fixed

jnewbery committed:
[wallet] Refactor ProcessImport()

This commit is move-only and doesn't make any functional changes. It
simply moves code around within ProcessImport() in preparation for
refactors in the next commits.

jnewbery committed:
[wallet] Add ProcessImportLegacy()

This commit adds a ProcessImportLegacy() function which
currently does nothing. It also unindents a block of
code for a future move-only change.

Reviewer hint: review with -w to ignore whitespace changes.

jnewbery committed:
[wallet] Refactor ProcessImport() to call ProcessImportLegacy()

This is almost entirely a move-only commit.

Reviewer hint: use --color-moved=zebra for review.

jnewbery commented:
I've had a go at rearranging the commits for this PR in a branch here: https://github.com/jnewbery/bitcoin/tree/pr14491.rearrange. It splits up the code move and new functionality, so with `-w` and `--color-moved=zebra` for the move commits, it should be a lot easier to review (review hints are in the commit log). I've also made a very minor change that undoes the reversion here: https://github.com/bitcoin/bitcoin/pull/14491#discussion_r247184235.

I think splitting up the commits in this way makes the PR much easier to review. @MeshCollider - feel free to grab that branch if you agree.

Sjors commented:
@jnewbery thanks, that is indeed way more readable. For those looking on Github, you can use `?w=1`. So that just leaves my request for more tests.

I'm quite happy with how short the meat of this change in https://github.com/bitcoin/bitcoin/commit/e7c45d7eb425fbe8c7b9f92d21d10c99145d3771 is.

MeshCollider commented:
@jnewbery Thanks! I've grabbed your branch, and I'll address the few other comments shortly

jnewbery commented on src/wallet/rpcdump.cpp:
```diff
@@ -923,154 +924,256 @@ static std::string RecurseImportData(const CScript& script, ImportData& import_d
         // Check whether we have any work to do
-        if (::IsMine(*pwallet, script) & ISMINE_SPENDABLE) {
-            throw JSONRPCError(RPC_WALLET_ERROR, "The wallet already contains the private key for this address or script");
+        for (const CScript& script : script_pub_keys) {
```
I could imagine this being an annoying failure mode if you try to import a ranged descriptor where the range is over something you've already imported. The error message doesn't give any indication of which script_pub_key caused the import to fail.

MeshCollider replied:
Fixed

ryanofsky reviewed (commented):
Reviewed 087ba0734ee24438b2f495367c71fb257bb5b760 but not acking, since I haven't thought through the effects of the new `ProcessImportDescriptor` function yet, especially the `AddWatchOnly` pubkey stuff discussed in https://github.com/bitcoin/bitcoin/pull/14491#discussion_r247192015 and https://github.com/bitcoin/bitcoin/pull/14491#discussion_r225408327.

The bulk of changes in this PR look very good, though, and the review comments I left are all minor, so feel free to ignore them.

@MeshCollider it would be helpful you could followup on the two AddWatchOnly threads https://github.com/bitcoin/bitcoin/pull/14491#discussion_r247192015 https://github.com/bitcoin/bitcoin/pull/14491#discussion_r225408327 to say what the current status is.

MeshCollider commented on src/wallet/rpcdump.cpp:
```diff
@@ -942,9 +955,7 @@ static UniValue ProcessImport(CWallet * const pwallet, const UniValue& data, con
-        const std::string& label = data.exists("label") ? data["label"].get_str() : "";
 
         // Generate the script and destination for the scriptPubKey provided
         CScript script;
```
@ryanofsky `script` is used below, in the `RecurseImportData` call. I have reduced the scope of `dest` though.

MeshCollider commented on src/wallet/rpcdump.cpp:
```diff
@@ -1062,6 +1063,83 @@ static UniValue ProcessImportLegacy(ImportData& import_data, std::map<CKeyID, CP
+        if (!range.exists("end")) {
+            throw JSONRPCError(RPC_INVALID_PARAMETER, "End of range for descriptor must be specified");
+        }
+        range_end = range["end"].get_int64();
```
Done

MeshCollider commented on src/wallet/rpcdump.cpp:
```diff
@@ -1062,6 +1063,83 @@ static UniValue ProcessImportLegacy(ImportData& import_data, std::map<CKeyID, CP
+
+    int64_t range_start = 0, range_end = 0;
+    if (!parsed_desc->IsRange() && data.exists("range")) {
+        throw JSONRPCError(RPC_INVALID_PARAMETER, "Range should not be specified for an un-ranged descriptor");
```
Done

sipa commented:
I think you've now changed it to never importing public keys.

If I import "pkh(036f1aef8329e88a8e7dca56a4e8f908697555478b26709f7513eba54db4acea21)", and subsequently call `getaddressinfo 1H2JaLY37d3PeqGgMV6yDStzHqenoRcCYF`, it says "solvable: False". All information for solvability is available.

Sjors commented:
The following tests reproduces what @sipa found:

```python
        # Test importing of a P2PKH address via descriptor
        key = get_key(self.nodes[0])
        self.log.info("Should import a p2pkh address from descriptor")
        self.test_importmulti({"desc": "pkh(" + key.pubkey + ")",
                               "timestamp": "now",
                               "label": "Descriptor import test"},
                              True,
                              warnings=["Some private keys are missing, outputs will be considered watchonly. If this is intentional, specify the watchonly flag."])
        test_address(self.nodes[1],
                     key.p2pkh_addr,
                     solvable=True,
                     ismine=False,
                     label="Descriptor import test")
```

ryanofsky commented:
More IRC discussion: http://www.erisian.com.au/bitcoin-core-dev/log-2019-01-24.html#l-742

Beyond the immediate bug that's been found, I'm confused about the pubkey issues being discussed:
https://github.com/bitcoin/bitcoin/pull/14491#discussion_r247192015, https://github.com/bitcoin/bitcoin/pull/14491#discussion_r225408327. Just at a really high level it's not clear to me if there is:

1. One clearly correct solution to implement in this PR.
2. One solution to implement in this PR that is clearly better than alternatives, but with known limitations that can be addressed later.
3. Disagreement about multiple solutions with different tradeoffs.

@sipa / @MeshCollider if you could clarify whether the situation is (1), (2), or (3), it would help the discussion make more sense...

sipa commented:
@ryanofsky I suspect this is subjective, but this is my view:

* Due to the inherent quirks in the current IsMine logic, we can't guarantee that *just* the exact scriptPubKey implied by the provided descriptor will match. If that was possible, it would be the obvious solution - but it's a choice we make by implementing this as a layer on top of the existing IsMine logic.
* A strong requirement in my view is that all the things that are imported are "policy compatible" with the descriptor. By that I mean that all scriptPubKeys made watched by importing a descriptor must be spendable by the same group of keys that can spend the sPK corresponding to the descriptor. If this is violated, you risk accepting a payment to something you can't spend.
* A weaker requirement is that whenever all information for solvability is present in the descriptor, the output is considered solvable.

I believe the above is all possible with the current IsMine logic, so I think it's the most obvious solution. It requires only importing the public keys which occur inside P2PKH/P2WPKH, but not those inside multisig.

sipa commented:
@MeshCollider See #15263, that should make it easier.

jonasschnelli commented:
Tested a bit and works as expected (limited tests).
A nitpick would be why the range is a JSON object rather then an array with two items.

MeshCollider commented:
@jonasschnelli I had the array initially, but the RPCHelpMan doesn't support that kind of help output so I just changed it

kallewoof reviewed (commented):
utACK with some nits

Sjors commented:
I took a stab at rebasing this, but it was absolute hell :-) Maybe @achow101 has more luck...

achow101 commented:
https://github.com/achow101/bitcoin/tree/rebase-14491 is a rebase of this which handles the merge conflicts caused by #15235. It does not account for #15263 (even though that is in it's history).

MeshCollider commented:
I've rebased and squashed as well as added some more tests for descriptor imports and fixed the public key issue now #15263 has gone in

Sjors commented on src/wallet/rpcdump.cpp:
```diff
@@ -1103,6 +1108,91 @@ static UniValue ProcessImportLegacy(ImportData& import_data, std::map<CKeyID, CP
+    FlatSigningProvider out_keys;
+
+    // Expand all descriptors to get public keys and scripts.
+    // TODO: get private keys from descriptors too
```
Nit: add (TODO to add) `expandRange` method to Descriptor instance

MeshCollider replied:
@Sjors this TODO is addressed in #15024 :)

Sjors reviewed (approved):
utACK b985e9c

ryanofsky reviewed (approved):
utACK b985e9c850ea682eced7021faf6c7c835066c61b. Changes since last review: rebase, more tests, changes to comments, documentation, and error checking.
